### PR TITLE
feat: 🎸 Filter by url query param for seriesInstnaceUID

### PR DESCRIPTION
**Pull request by ladeirarodolfo**
### Includes
- Filter series by query param (works for default and gcloud). IT FILTERS ONLY SERIES (no sopInstance yet)
- Filter applies to lazy or not study loading
- Uses Snackbar to toast user messages.

#### Related
- There will be another version (PR to be opened soon), which will promote instead of filter series.

### Screnshoots
undefined-_2_
Untitled_-Oct-28_-2019-6_25-PM

### CODE REVIEW

**Comment by dannyrb** [comment]
@swederik is there a simple way for me to test this? If not currently, is there a simple/sane approach to make it possible?

File: platform/core/src/studies/retrieveStudiesMetadata.js [hunk 1]
```diff
@@ -13,19 +13,15 @@ import { retrieveStudyMetadata } from './retrieveStudyMetadata';
 export default function retrieveStudiesMetadata(
   server,
   studyInstanceUids,
-  seriesInstanceUids
+  filters
```
Comment (dannyrb) [inline platform/core/src/studies/retrieveStudiesMetadata.js]: Can we destructure this at some point, so it's clearer what possible filters can be?

File: platform/core/src/studies/retrieveStudyMetadata.js [hunk 2]
```diff
@@ -11,7 +11,7 @@ const StudyMetaDataPromises = new Map();
  * @param {string} studyInstanceUid The UID of the Study to be retrieved
  * @returns {Promise} that will be resolved with the metadata or rejected with the error
  */
-export function retrieveStudyMetadata(server, studyInstanceUid) {
+export function retrieveStudyMetadata(server, studyInstanceUid, filters) {
```
Comment (dannyrb) [inline platform/core/src/studies/retrieveStudyMetadata.js]: Can we update the JSDoc and destructure here so it's clearer which filters can be passed?

File: platform/core/src/studies/services/wado/retrieveMetadata.js [hunk 3]
```diff
@@ -468,37 +468,70 @@ async function createStudyFromSOPInstanceList(server, sopInstanceList) {
  * @param {string} studyInstanceUid The Study Instance UID of the study which needs to be loaded
  * @returns {Object} A study descriptor object
  */
-async function RetrieveMetadata(server, studyInstanceUid) {
-  return (server.enableStudyLazyLoad !== false
+async function RetrieveMetadata(server, studyInstanceUid, filters) {
+  return (server.enableStudyLazyLoad !== false && false
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadata.js]: `&& false` makes this always `false`?

File: platform/core/src/studies/services/wado/retrieveMetadata.js [hunk 4]
```diff
@@ -468,37 +468,70 @@ async function createStudyFromSOPInstanceList(server, sopInstanceList) {
  * @param {string} studyInstanceUid The Study Instance UID of the study which needs to be loaded
  * @returns {Object} A study descriptor object
  */
-async function RetrieveMetadata(server, studyInstanceUid) {
-  return (server.enableStudyLazyLoad !== false
+async function RetrieveMetadata(server, studyInstanceUid, filters) {
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadata.js]: Update `JSDoc`, and what can be in `filters`?

File: platform/core/src/studies/services/wado/retrieveMetadata.js [hunk 5]
```diff
@@ -468,37 +468,70 @@ async function createStudyFromSOPInstanceList(server, sopInstanceList) {
...
 }
 
 /**
  *
  * @param {*} server
  * @param {*} studyInstanceUID
  */
-async function loadStudyMetadata(server, studyInstanceUID) {
-  const dicomWeb = new api.DICOMwebClient({
+async function loadStudyMetadata(server, studyInstanceUID, filters) {
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadata.js]: Update JSDOC

If it feels like I'm asking "what can be in `filters`"? too much, then that might be indicative of two things:

1. Why are we passing filters so far down the callstack? Does every layer of this beast have a good purpose? Does each actually need to know our filters?
2. Is there a better pattern for passing complex values?

File: platform/core/src/studies/services/wado/retrieveMetadata.js [hunk 6]
```diff
@@ -468,37 +468,70 @@ async function createStudyFromSOPInstanceList(server, sopInstanceList) {
...
   });
-  return dicomWeb
-    .retrieveStudyMetadata({ studyInstanceUID })
-    .then(result => createStudyFromSOPInstanceList(server, result));
+
+  const options = {
+    studyInstanceUID,
+  };
+
+  const { seriesInstanceUID } = filters;
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadata.js]: Ah. Here it is :Wave:

File: platform/viewer/src/routes/ViewerRouting.js [hunk 7]
```diff
@@ -36,7 +36,7 @@ function ViewerRouting({ match: routeMatch, location: routeLocation }) {
     return (
       <ConnectedViewerRetrieveStudyData
         studyInstanceUids={studyUIDs}
-        seriesInstanceUids={seriesUIDs}
+        seriesInstanceUIDs={seriesUIDs}
```
Comment (dannyrb) [inline platform/viewer/src/routes/ViewerRouting.js]: Why the change in casing? I feel like we more consistently use `Uids`.

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js [hunk 8]
```diff
@@ -118,4 +155,4 @@ class ViewerRetrieveStudyData extends Component {
   }
 }
 
-export default ViewerRetrieveStudyData;
+export default withSnackbar(ViewerRetrieveStudyData);
```
Comment (dannyrb) [inline platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: This is a pretty simple component. We could just make it a functional component?

File: platform/core/src/studies/services/wado/retrieveMetadata.js [hunk 9]
```diff
@@ -468,37 +468,70 @@ async function createStudyFromSOPInstanceList(server, sopInstanceList) {
...
     url: server.wadoRoot,
     headers: DICOMWeb.getAuthorizationHeader(server),
   });
+
   const seriesLoader = makeSeriesLoader(
     dicomWeb,
-    studyInstanceUid,
-    seriesInstanceUids
+    studyInstanceUID,
+    seriesInstanceUIDsMap
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadata.js]: casing change?

File: platform/core/src/studies/services/wado/retrieveMetadata.js [hunk 10]
```diff
@@ -468,37 +468,70 @@ async function createStudyFromSOPInstanceList(server, sopInstanceList) {
...
+  try {
+    studyMetadata = await retrieveMethod.call(client, options);
+  } catch (e) {
+    // in case occurred error when retrieveSeriesMetadata, retrieve again from retrieveStudyMetadata
+    if (seriesInstanceUID) {
+      studyMetadata = await client.retrieveStudyMetadata({ studyInstanceUID });
+    } else {
+      throw e;
+    }
+  }
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadata.js]: Wondering if we should clean this up a little to make it clearer what's going on?

0. Identify request + options
1. Primary request
2. Fallback request
3. ???
4. Profit

File: platform/core/src/studies/services/wado/retrieveMetadata.js (hunk 5 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/services/wado/retrieveMetadata.js]: Definitely we can revisit and discuss about these many layers.. I dont see a purpose of having that many.
"filters" must be passed to the very end layer which does the call to retrieve series (in case of filter), because lazy load uses the first one (in our case would be the filtered) to trigger the process.

There is another logic which  we can place the filtering/promoting at the very beginning layer, but from my understanding this wouldnt be a good user experience at all (page will full load, than filter will be applied). I.e on a lazy env.. studies will be listed then filtered.

Let me know what you think?

File: platform/core/src/studies/retrieveStudiesMetadata.js (hunk 1 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/retrieveStudiesMetadata.js]: ok

File: platform/core/src/studies/services/wado/retrieveMetadata.js (hunk 3 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/services/wado/retrieveMetadata.js]: debugging flag :D removing soon

File: platform/core/src/studies/services/wado/retrieveMetadata.js (hunk 4 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/services/wado/retrieveMetadata.js]: ok

File: platform/core/src/studies/services/wado/retrieveMetadata.js (hunk 10 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/services/wado/retrieveMetadata.js]: ok changed a bit the approach

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js (hunk 8 above)
Comment (ladeirarodolfo) [reply platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: yes.. sure.. can we cover on a separate PR?

File: platform/core/src/studies/services/wado/retrieveMetadata.js (hunk 6 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/services/wado/retrieveMetadata.js]: :D updated JSDOCs

File: platform/viewer/src/routes/ViewerRouting.js (hunk 7 above)
Comment (ladeirarodolfo) [reply platform/viewer/src/routes/ViewerRouting.js]: I spoke to Erik and he mentioned UID would be more appropriate.
But again.. I see places with different versions.. I`m not that confident using one or other..
totally opened to implement what you think is the best one

Comment (dannyrb) [reply platform/viewer/src/routes/ViewerRouting.js]: I don't have a strong preference, I just want us to be consistent. I'll poke Erik tomorrow.

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js [hunk 11]
```diff
@@ -12,7 +13,7 @@ const { studyMetadataManager, updateMetaDataManager } = utils;
 class ViewerRetrieveStudyData extends Component {
   static propTypes = {
     studyInstanceUids: PropTypes.array.isRequired,
-    seriesInstanceUids: PropTypes.array,
+    seriesInstanceUIDs: PropTypes.array,
```
Comment (dannyrb) [inline platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: It's still confusing/odd to me that we're keeping `studyInstanceUids` but changing the casing of `seriesInstanceUIDs`. Unless we're changing the usage throughout, I'd rather we avoid doing this. At a minimum, we should use the same casing rules within the same functions/scope.

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js [hunk 12]
```diff
@@ -26,19 +27,55 @@ class ViewerRetrieveStudyData extends Component {
 
   async loadStudies() {
     try {
-      const { server, studyInstanceUids, seriesInstanceUids } = this.props;
+      const { server, studyInstanceUids, seriesInstanceUIDs } = this.props;
+      const filters = {};
+
+      // Use the first, discard others
+      const seriesInstanceUID = seriesInstanceUIDs && seriesInstanceUIDs[0];
```
Comment (dannyrb) [inline platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: Where `seriesInstanceUIDs` is set by the route?

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js [hunk 13]
```diff
@@ -26,19 +27,55 @@ class ViewerRetrieveStudyData extends Component {
...
+    // skip in case no filter or no toast manager
+    if (!seriesInstanceUID || !snackbarContext) {
+      return;
+    }
+
+    const firstStudy = studies[0] || {};
+    const { seriesList = [] } = firstStudy;
+    const firstSeries = seriesList[0];
+
+    if (!firstSeries || firstSeries.seriesInstanceUid !== seriesInstanceUID) {
```
Comment (dannyrb) [inline platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: Our "filter validation" is really a test to see that _any_ results were returned after we've applied our filters to our query, right? It might be easier to just check that we returned a study with one or more series?

File: platform/core/src/studies/services/wado/retrieveMetadata.js [hunk 14]
```diff
@@ -1,609 +1,30 @@
...
-  if (Array.isArray(sopInstanceList) && sopInstanceList.length > 0) {
-    const firstSopInstance = sopInstanceList[0];
-    const study = createStudy(server, firstSopInstance);
-    await addInstancesToStudy(server, study, sopInstanceList);
-    return study;
-  }
-  throw new Error('Failed to create study out of provided SOP instance list');
-}
+import RetrieveMetadataLoaderSync from './retrieveMetadataLoaderSync';
+import RetrieveMetadataLoaderAsync from './retrieveMetadataLoaderAsync';
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadata.js]: Ah, okay. So just splitting up this logic to isolate it a bit.

File: platform/core/src/studies/services/wado/retrieveMetadataLoader.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
+/**
+ * Class to define inheritance of load retrieve strategy.
+ * The process can be async load (lazy) or sync load
+ *
+ * There are methods that must be implemented at consumer level
+ * To retrieve study call execLoad
+ */
+export default class RetrieveMetadataLoader {
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadataLoader.js]: 👋 Hello Mr. Abstract Class

File: platform/core/src/studies/studyUtils.js [hunk 16]
```diff
@@ -0,0 +1,413 @@
+import DICOMWeb from '../DICOMWeb/';
```
Comment (dannyrb) [inline platform/core/src/studies/studyUtils.js]: If this is the shared logic for the metadata loaders, would it make more sense as "internals" that live in the same location next to those files?

Comment (dannyrb) [reply platform/core/src/studies/studyUtils.js]: Naming files "utilities" is a bit dangerous, in my opinion. I often think of utilities as general purpose static methods. These feel like data mappings, and a few methods for parsing data, building URLs, etc.

If it's simply shared logic you've tried to factor out, you can name it that, and put it close to the files that use it (if they're co-located). If it's generic data "mapping", then maybe we need to have a discussion (as a team) about creating a "mapping" layer where we discuss a more formal "Aggregate Root", "Viewmodel", and "Data Transfer Object" setup.

File: platform/core/src/studies/services/wado/retrieveMetadataLoaderAsync.js [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+import * as StudyUtils from '../../studyUtils';
+
+/**
+ * Search series of a given study
+ * @param {DICOMwebClient} client Dicomweb client api
+ * @param {string} studyInstanceUID The Study Instance UID to search series from;
+ * @returns {Arrays} A list of Series Instances
+ */
+async function searchStudySeries(client, studyInstanceUID) {
+  const seriesList = await client.searchForSeries({ studyInstanceUID });
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadataLoaderAsync.js]: What data is fetched from this? Is there anyway to apply our series specific filter with this request?

File: platform/core/src/studies/services/wado/retrieveMetadataLoaderAsync.js [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+
+/**
+ * Filter seriesList
+ * @param {Arrays} seriesList list of Series Instance UIDs
+ * @param {Object} filters object containing filter to be applied.
+ * @returns {Arrays} A list of Series Instances
+ */
+function filterStudySeries(seriesList = [], filters) {
+  let result = [...seriesList];
+  const { seriesInstanceUID: toCompare } = filters;
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadataLoaderAsync.js]: So at this point, we at least have _some_ metadata for _every_ series in our study. I know in our `studyList` search, we're able to apply filters to fields. Can we apply a filter to the seriesInstanceUid field when we query?

See: https://www.dicomstandard.org/dicomweb/query-qido-rs/
First `key`/`value` matching

Once we have a list of all the `seriesInstanceUids` that qualify for our study + filter, we can then do the heavier metadata and instance calls? Also lets us skip filtering on the client (less data transmission)

File: platform/core/src/studies/services/wado/retrieveMetadataLoaderSync.js [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+   */
+  *getLoaders() {
+    const loaders = [];
+    const { filters: { seriesInstanceUID } = {}, client } = this;
+
+    if (seriesInstanceUID) {
+      loaders.push(client.retrieveSeriesMetadata);
+    }
+
+    loaders.push(client.retrieveStudyMetadata);
```
Comment (dannyrb) [inline platform/core/src/studies/services/wado/retrieveMetadataLoaderSync.js]: I think I see where/how we're ultimately filtering our result set in the `async` implementation. I'm pretty sure this is how/where we accomplish this here. I'm not 100% sure I understand why it works here. Can you give me a high level overview?

File: platform/core/src/studies/getSeriesInfo.js [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+    info = Object.freeze({
+      modality,
+      isLowPriority: isLowPriorityModality(modality),
+      seriesInstanceUid: DICOMWeb.getString(series['0020000E']),
+      seriesNumber: DICOMWeb.getNumber(series['00200011'], 0) || 0,
+    });
+    series[INFO] = info;
+  }
+  return info;
+}
```
Comment (dannyrb) [inline platform/core/src/studies/getSeriesInfo.js]: Is this new code? What are we using this for? Is this a common approach for mapping/persisting reusable models?

Is there a better, more descriptive name, than `getSeriesInfo`?

File: platform/core/src/studies/sortStudy.js [hunk 21]
```diff
@@ -1,17 +1,96 @@
+import getSeriesInfo from './getSeriesInfo';
```
Comment (dannyrb) [inline platform/core/src/studies/sortStudy.js]: Can you walk me through the changes in this file? It looks like we added a fair amount of code here, and I want to understand the benefits + usage

File: platform/viewer/src/routes/ViewerRouting.js (hunk 7 above)
Comment (ladeirarodolfo) [reply platform/viewer/src/routes/ViewerRouting.js]: I will change it to seriesInstanceUids!

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js (hunk 12 above)
Comment (ladeirarodolfo) [reply platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: or from ViewerRouting or from useLocation hooks. I`m using the first strategy because it was already present on the code.

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js (hunk 13 above)
Comment (ladeirarodolfo) [reply platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: not actually because as per requirement, it will always return series (filtered or not)..

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js (hunk 11 above)
Comment (ladeirarodolfo) [reply platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: changed to seriesInstanceUids

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js (hunk 13 above)
Comment (dannyrb) [reply platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: Ah. Okay. So this is further down the pipe. We've hit our fallback (study retrieve), so we're more or less guaranteed to have data. We have to check the prop to see if filtering was applied.

What if the first series in the list just happens to match our filter?

File: platform/core/src/studies/studyUtils.js (hunk 16 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/studyUtils.js]: ok changed it a bit..

File: platform/core/src/studies/getSeriesInfo.js (hunk 20 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/getSeriesInfo.js]: this is not new code.. just moved to a shared place where sortStudy and retrieveMetadaAsync can use...

File: platform/core/src/studies/sortStudy.js (hunk 21 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/sortStudy.js]: actually there is nothing new.. I preserved old implementation of sortStudy and exposed what has being used by retrieveMetadaAsync file.. but its always a pleasure to walk you through :D

File: platform/core/src/studies/services/wado/retrieveMetadataLoaderSync.js (hunk 19 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/services/wado/retrieveMetadataLoaderSync.js]: yes.. lets do it..

File: platform/core/src/studies/services/wado/retrieveMetadataLoaderAsync.js (hunk 17 above)
Comment (ladeirarodolfo) [reply platform/core/src/studies/services/wado/retrieveMetadataLoaderAsync.js]: ok changed it a bit.. using api to fetch series (filtered)

**Comment by ladeirarodolfo** [comment]
To test it:

Filter
Open any study with valid query param ?SeriesInstanceUID
https://viewer.ohif.org/viewer/1.3.6.1.4.1.25403.345050719074.3824.20170126085406.1?SeriesInstanceUID=2.25.856992923281239003733479634608709215918

Not Filter
Open any study without query param ?SeriesInstanceUID

Fail scenarios
Open any study with invalid query param ?SeriesInstanceUID


Studies
Study 1
https://viewer.ohif.org/viewer/1.3.6.1.4.1.25403.345050719074.3824.20170126085406.1

Valid SeriesInstanceUID values
2.25.856992923281239003733479634608709215918
2.25.374532628115939034791691984300357809891
2.25.260236938056495981644704674705700826124
2.25.216660299853408893926642702413523431206
2.25.810301644129899623236282516055099191586
1.3.6.1.4.1.25403.345050719074.3824.20170126085406.5


Study 2
https://viewer.ohif.org/viewer/1.2.840.113619.2.30.1.1762295590.1623.978668949.886

Valid SeriesInstanceUID values
1.2.840.113619.2.30.1.1762295590.1623.978668949.890
1.2.840.113619.2.30.1.1762295590.1623.978668949.887
2.25.362578567227550924517643184276356001107
2.25.930912967783862238320235018609232362156
2.25.335260806496828400742633309031250629174


How to get a valid seriesInstanceUID for given study?? (valid in case lazy load only. env has set it as true by default)
- Go to study list
- Open any study
- Go to viewer page
- Open devtools/network tab. Look for request `/studies/{studyInstanceUID}/series`
- Each item will be a series; Get related seriesInstanceUID  `Value` on key `0020000E`
